fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt

```diff
@@ -79,13 +104,76 @@ class WebExtensionPromptFeature(
 
     private fun handlePermissionRequest(
         addon: Addon,
-        promptRequest: WebExtensionPromptRequest.Permissions,
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions,
     ) {
-        if (hasExistingPermissionDialogFragment()) {
-            return
+        if (hasExistingPermissionDialogFragment()) return
+        showPermissionDialog(
+            addon,
+            promptRequest,
+        )
+    }
+
+    @VisibleForTesting
+    internal fun handleInstallationFailedRequest(
+        exception: WebExtensionInstallException,
+    ) {
+        val addonName = exception.extensionName ?: ""
+        var title = context.getString(R.string.mozac_feature_addons_failed_to_install, "")
+        val message = when (exception) {
+            is WebExtensionInstallException.Blocklisted -> {
+                context.getString(R.string.mozac_feature_addons_blocklisted, addonName)
+            }
+
+            is WebExtensionInstallException.Unknown, is WebExtensionInstallException.UserCancelled -> {
+                // Making sure we don't have a
+                // Title = Failed to install
+                // Message = Failed to install $addonName
+                title = ""
+                context.getString(R.string.mozac_feature_addons_failed_to_install, addonName)
+            }
+
+            is WebExtensionInstallException.NetworkFailure -> {
+                context.getString(R.string.mozac_feature_addons_failed_to_install_network_error)
+            }
+
+            is WebExtensionInstallException.CorruptedFile -> {
+                context.getString(R.string.mozac_feature_addons_failed_to_install_corrupt_error)
+            }
+
+            is WebExtensionInstallException.IncorrectHash -> {
+                context.getString(R.string.mozac_feature_addons_failed_to_install_hash_error)
+            }
+
+            is WebExtensionInstallException.FileAccess -> {
+                context.getString(
+                    R.string.mozac_feature_addons_failed_to_install_file_access_error,
+                    addonName,
+                )
+            }
+
+            is WebExtensionInstallException.UnSigned -> {
+                context.getString(
+                    R.string.mozac_feature_addons_failed_to_install_un_signed_error,
+                    addonName,
+                )
+            }
+
+            is WebExtensionInstallException.Incompatible -> {
+                val appName = context.getString(R.string.app_name)
+                val version = context.appVersionName
+                context.getString(
+                    R.string.mozac_feature_addons_failed_to_install_in_compatible_error,
+                    addonName,
+                    appName,
+                    version,
+                )
+            }
```

Comment:
Shouldn't we have a default case?

Reply:
Do you mean for the `when` expression, like an `else` statement ?

Reply:
If so, it's not needed as we are using a [sealed class](https://kotlinlang.org/docs/sealed-classes.html#sealed-classes-and-when-expression) which in a [when expression](https://kotlinlang.org/docs/sealed-classes.html#sealed-classes-and-when-expression) enforce by the compiler that all cases are fulfilled.